### PR TITLE
Assessment Report

### PR DESCRIPTION
**Pull request by priyaannu**
Re-designed the Assessment Report  

- summary of the schema conversion
- summarized table report by aggregating issues by issue type
- display table numbers beside the issue description
- display table names as an expansion body of the issue description 
- made a typeDescription to map to typeEnum of structured report
- display only those issue types that are present in the user's schema report

Made some changes in structured report format
- Replaced warning by issue 
- Added typeEnum and description field in structured report under issuesList
- Added "isSharded" field in structured report
- Corrected the calculation of number of warnings in "warnings" field in structured report
- Added "issues" field in structured report's table report which gives total number of issues in that table

Tests

- updated the mySQL and postgre report_test files
- tested manually for possible conditions

### CODE REVIEW

File: internal/reports/report_v2.go [hunk 1]
```diff
@@ -60,15 +60,20 @@ type NameChange struct {
...
 
-type Warnings struct {
-	WarningType string   `json:"warningType"`
-	WarningList []string `json:"warningList"`
+type Issues struct {
+	IssueType string            `json:"issueType"`
+	IssueList []IssueClassified `json:"issueList"`
+}
+
+type IssueClassified struct {
```
Comment (shreyakhajanchi) [inline internal/reports/report_v2.go:68]: Lets rename it to IssueType and variables inside it as Type & Description

File: webv2/summary/summary_service.go [hunk 2]
```diff
@@ -8,39 +8,39 @@ import (
 // getSummary returns table wise summary of conversion.
 func getSummary() map[string]ConversionSummary {
 	sessionState := session.GetSessionState()
-	reports := reports.AnalyzeTables(sessionState.Conv, nil)
+	Reports := reports.AnalyzeTables(sessionState.Conv, nil)
```
Comment (shreyakhajanchi) [inline webv2/summary/summary_service.go:11]: Why do we need Reports to be public?

File: ui/src/app/model/structured-report.ts [hunk 3]
```diff
@@ -47,9 +48,14 @@ export interface INameChange {
...
 
-export interface IWarnings {
-	warningType: string
-	warningList: string[]
+export interface IIssues {
+	issueType: string
+	issueList: IssueClassified[]
+}
+
+export interface IssueClassified {
```
Comment (shreyakhajanchi) [inline ui/src/app/model/structured-report.ts:56]: same renaming here

File: ui/src/app/model/issue-report.ts [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    suggestions: Map<string, IssueDescription>
+    notes: Map<string, IssueDescription>
+}
+
+export interface IssueDescription {
+    tableCount: number
+    tableNames: Set<string>
+}
+
+export const TypeDescription = {
```
Comment (shreyakhajanchi) [inline ui/src/app/model/issue-report.ts:13]: Lets get this value from backend via an API to avoid error scenarios

File: ui/src/app/components/sidenav-view-assessment/sidenav-view-assessment.component.scss [hunk 5]
```diff
@@ -35,30 +35,177 @@ table.mat-table {
...
+        margin-left: 1.25px;
+      }
     }
   }
+
   .content {
     padding-left: 15px;
     padding-top: 10px;
+
+    .sidenav-percentage-indent {
```
Comment (shreyakhajanchi) [inline ui/src/app/components/sidenav-view-assessment/sidenav-view-assessment.component.scss:66]: Move css to styles.css as much as possible

File: ui/src/app/components/sidenav-view-assessment/sidenav-view-assessment.component.html [hunk 6]
```diff
@@ -70,49 +64,249 @@ <h3 class="sidenav-title">Conversion status by table</h3>
...
+          <tr mat-row *matRowDef="let element; columns: columnsToDisplayWithExpand;" class="example-element-row"
+            [class.example-expanded-row]="isRowExpanded(element)" (click)="toggleRow(element)"></tr>
+          <tr mat-row *matRowDef="let row; columns: ['expandedDetail']" class="example-detail-row"></tr>
+        </table>
+      </mat-expansion-panel>
+      <br *ngIf="issueTableData_Errors.length">
+      <mat-expansion-panel *ngIf="issueTableData_Warnings.length">
+        <mat-expansion-panel-header>
+          <mat-panel-title>
+            <mat-icon matTooltip="Warning : Changes made because of differences in source and spanner capabilities."
```
Comment (shreyakhajanchi) [inline ui/src/app/components/sidenav-view-assessment/sidenav-view-assessment.component.html:138]: Let's reword it to: Warning : Schema modifications owing to difference in source and spanner capabilities.
Thoughts: @darshan-sj

File: ui/src/app/components/sidenav-view-assessment/sidenav-view-assessment.component.html [hunk 7]
```diff
@@ -70,49 +64,249 @@ <h3 class="sidenav-title">Conversion status by table</h3>
...
+            [class.example-expanded-row]="isRowExpanded(element)" (click)="toggleRow(element)"></tr>
+          <tr mat-row *matRowDef="let row; columns: ['expandedDetail']" class="example-detail-row"></tr>
+        </table>
+      </mat-expansion-panel>
+      <br *ngIf="issueTableData_Warnings.length">
+      <mat-expansion-panel *ngIf="issueTableData_Suggestions.length">
+        <mat-expansion-panel-header>
+          <mat-panel-title>
+            <mat-icon
+              matTooltip="Suggestion : We highly recommend you make these changes or else it will impact your DB performance."
```
Comment (shreyakhajanchi) [inline ui/src/app/components/sidenav-view-assessment/sidenav-view-assessment.component.html:192]: Suggestion: We recommend you make these changes to improve your DB performance.
cc: @darshan-sj

File: ui/src/app/components/sidenav-view-assessment/sidenav-view-assessment.component.html [hunk 8]
```diff
@@ -70,49 +64,249 @@ <h3 class="sidenav-title">Conversion status by table</h3>
...
-        >
-        <span><mat-icon class="icon success"> circle </mat-icon> <span>Converted automatically</span></span>
-      </div>
-    </mat-card>
+    <h3>Summarized Table Report</h3>
+    <div class="issue-report">
+      <mat-expansion-panel *ngIf="issueTableData_Errors.length">
+        <mat-expansion-panel-header>
+          <mat-panel-title>
+            <mat-icon matTooltip="Error: Please resolve them to proceed with the migration" matTooltipPosition="above"
```
Comment (shreyakhajanchi) [inline ui/src/app/components/sidenav-view-assessment/sidenav-view-assessment.component.html:86]: Error: Please resolve all the errors to proceed with the migration

File: webv2/summary/summary_service.go (hunk 2 above)
Comment (priyaannu) [reply webv2/summary/summary_service.go:11]: Yes we don't need it to be public, corrected!

File: internal/reports/report_v2.go [hunk 9]
```diff
@@ -60,15 +60,21 @@ type NameChange struct {
...
+	Category    string `json:"category"`
+	Description string `json:"description"`
 }
 
 type SchemaReport struct {
 	Rating       string `json:"rating"`
 	PkMissing    bool   `json:"pkMissing"`
-	Warnings     int64  `json:"warnings"`
+	Issues       int64  `json:"issues"`
+	Warnings 	 int64  `json:"warnings"`
```
Comment (shreyakhajanchi) [inline internal/reports/report_v2.go:77]: apply formatting

File: test_data/postgres_structured_report.json [hunk 10]
```diff
@@ -1,155 +1,185 @@
...
+                  {
+                     "category": "STORAGE_WARNING",
+                     "description": "Table bad_schema: Some columns will consume more storage in Spanner e.g. for column 'b', source DB type int4 is mapped to Spanner data type int64"
+                  },
+                  {
+                     "category": "MULTI_DIMENSIONAL_ARRAY_USES",
+                     "description": "Table 'bad_schema': Column 'c', type int4[4][2] is mapped to string(max). Spanner doesn't support multi-dimensional arrays"
+                  },
+                  {
+                     "category": "INAPPROPIATE_TYPE",
```
Comment (shreyakhajanchi) [inline test_data/postgres_structured_report.json:68]: correct the spelling - inappropriate

File: ui/src/app/model/issue-report.ts [hunk 11]
```diff
@@ -0,0 +1,12 @@
+export default interface IIssueReport {
+    errors: Map<string, IssueDescription>
+    warnings: Map<string, IssueDescription>
+    suggestions: Map<string, IssueDescription>
+    notes: Map<string, IssueDescription>
+}
+
+export interface IssueDescription {
```
Comment (shreyakhajanchi) [inline ui/src/app/model/issue-report.ts:8]: This interface doesn't contain issue description it only contains table names and table counts, rename appropriately

File: ui/src/app/model/summary.ts [hunk 12]
```diff
@@ -15,3 +15,8 @@ export interface ISummaryRow {
   content: string
   isRead: boolean
 }
+
+export interface Issue {
```
Comment (shreyakhajanchi) [inline ui/src/app/model/summary.ts:19]: Do not define this interface twice, use it from structured report

File: webv2/routes.go [hunk 13]
```diff
@@ -82,6 +82,9 @@ func getRoutes() *mux.Router {
 	// Summary
 	router.HandleFunc("/summary", summary.GetSummary).Methods("GET")
 
+	// Type Description
+	router.HandleFunc("/typeDescription", getTypeDescription).Methods("GET")
```
Comment (shreyakhajanchi) [inline webv2/routes.go:86]: rename it to 'getIssueDescription' both routes and method

File: webv2/web.go [hunk 14]
```diff
@@ -1402,6 +1402,21 @@ func getDSpannerDDL(w http.ResponseWriter, r *http.Request) {
 	json.NewEncoder(w).Encode(strings.Join(l, ""))
 }
 
+// getTypeDescription maps IssueDB's Category to corresponding CategoryDescription(if present),
+// or to the Brief if not present and pass the map to frontend to be used in assessment report UI
+func getTypeDescription(w http.ResponseWriter, r *http.Request) {
+	var CategoryMap = make(map[string]string)
```
Comment (shreyakhajanchi) [inline webv2/web.go:1408]: why is CategoryMap public , i.e. in caps?

Comment (shreyakhajanchi) [reply webv2/web.go:1408]: categorymap is very misleading , use IssuesMap or Issue Description

File: webv2/web.go [hunk 15]
```diff
@@ -1402,6 +1402,21 @@ func getDSpannerDDL(w http.ResponseWriter, r *http.Request) {
...
+	for _, varStruct := range reports.IssueDB {
```
Comment (shreyakhajanchi) [inline webv2/web.go:1409]: rename varStruct to issue

File: ui/src/app/components/sidenav-view-assessment/sidenav-view-assessment.component.ts [hunk 16]
```diff
@@ -6,30 +6,64 @@ import { SidenavService } from 'src/app/services/sidenav/sidenav.service'
...
+  description: string
+  tableCount: number
+  tableNamesJoinedByComma: string
+}
 
 @Component({
   selector: 'app-sidenav-view-assessment',
   templateUrl: './sidenav-view-assessment.component.html',
   styleUrls: ['./sidenav-view-assessment.component.scss'],
+  animations: [
```
Comment (shreyakhajanchi) [inline ui/src/app/components/sidenav-view-assessment/sidenav-view-assessment.component.ts:23]: can this not be moved to css ?

File: ui/src/app/components/sidenav-view-assessment/sidenav-view-assessment.component.ts [hunk 17]
```diff
@@ -134,4 +168,235 @@ export class SidenavViewAssessmentComponent implements OnInit {
...
+            let fetchedDescription = this.typeDescription[key]
+            this.issueTableData_Warnings.push({
+              position: i,
+              description: fetchedDescription,
+              tableCount: value.tableCount,
+              tableNamesJoinedByComma: tableNamesList.join(', '),
+            })
+            i += 1;
+          }
+        }
```
Comment (shreyakhajanchi) [inline ui/src/app/components/sidenav-view-assessment/sidenav-view-assessment.component.ts:342]: move this to a method and remove code duplication for four issue types

File: ui/src/app/components/sidenav-view-assessment/sidenav-view-assessment.component.ts [hunk 18]
```diff
@@ -134,4 +168,235 @@ export class SidenavViewAssessmentComponent implements OnInit {
...
+                    }
+                    descNew.tableNames.add(fetchedTableReport.srcTableName)
+                    descNew.tableCount = descNew.tableNames.size
+                    report.errors.set(errorIssue.category, descNew)
+                  } else {
+                    // if the error is new we initialise issue description and add the table to it
+                    let desc = defaultIssue
+                    desc.tableNames.add(fetchedTableReport.srcTableName)
+                    desc.tableCount = desc.tableNames.size
+                    report.errors.set(errorIssue.category, desc)
```
Comment (shreyakhajanchi) [inline ui/src/app/components/sidenav-view-assessment/sidenav-view-assessment.component.ts:231]: move this to a method and remove code duplication for 4 issue types

File: internal/reports/report_helpers.go [hunk 19]
```diff
@@ -42,8 +42,8 @@ type tableReport struct {
 }
 
 type tableReportBody struct {
-	Heading string
-	Lines   []string
+	Heading   string
+	IssueBody []Issue
 }
```
Comment (shreyakhajanchi) [inline internal/reports/report_helpers.go:48]: The discrepancy in issue number vs older warning count is coming because we are not taking into account the number of notes & suggestions similar to warnings and error

**Review by shreyakhajanchi — COMMENTED** [review]
Left some comments

File: webv2/web.go (hunk 14 above)
Comment (priyaannu) [reply webv2/web.go:1408]: Corrected it to issuesMap

File: internal/reports/report_helpers.go (hunk 19 above)
Comment (priyaannu) [reply internal/reports/report_helpers.go:48]: The issue number is getting calculated by iterating the Issues List, the discrepancy is in the warning count.
Resolving this for now, will discuss and commit the required changes further.

File: internal/reports/report_helpers.go [hunk 20]
```diff
@@ -383,40 +507,53 @@ func fillRowStats(conv *internal.Conv, srcTable string, badWrites map[string]int
...
+	internal.StringOverflow:       {Brief: "String overflow issue might occur as maximum supported length in Spanner is 2621440", severity: warning, Category: "STRING_OVERFLOW_WARNING"},
+	internal.HotspotTimestamp:     {Brief: "Timestamp Hotspot Occured", severity: warning, Category: "TIMESTAMP_HOTSPOT"},
+	internal.HotspotAutoIncrement: {Brief: "Autoincrement Hotspot Occured", severity: warning, Category: "AUTOINCREMENT_HOTSPOT"},
+	internal.InterleavedOrder: {Brief: "can be converted as Interleaved with Table", severity: suggestion, Category: "INTERLEAVE_TABLE_SUGGESTION",
+		CategoryDescription: "Some tables can be interleaved"},
+	internal.RedundantIndex:     {Brief: "Redundant Index", severity: warning, Category: "REDUNDANT_INDEX"},
+	internal.AutoIncrementIndex: {Brief: "Auto increment column in Index can create a Hotspot", severity: warning, Category: "AUTO-INCREMENT_INDEX"},
+	internal.InterleaveIndex: {Brief: "can be converted to an Interleave Index", severity: suggestion, Category: "INTERLEAVE_INDEX_SUGGESTION",
+		CategoryDescription: "Some columns can be interleaved"},
```
Comment (shreyakhajanchi) [inline internal/reports/report_helpers.go:539]: Rephrase the category description: Some columns can be converted to interleave index

File: internal/reports/report_helpers.go [hunk 21]
```diff
@@ -383,40 +507,53 @@ func fillRowStats(conv *internal.Conv, srcTable string, badWrites map[string]int
...
+	internal.IllegalName: {Brief: "Names must adhere to the spanner regular expression {a-z|A-Z}[{a-z|A-Z|0-9|_}+]", severity: warning, Category: "ILLEGAL_NAME"},
+	internal.InterleavedRenameColumn: {Brief: "Candidate for Interleaved Table", severity: suggestion, Category: "RENAME_INTERLEAVED_COLUMN_PRIMARY_KEY",
+		CategoryDescription: "If primary key is renamed in table to match the foreign key, the table can be interleaved"},
+	internal.InterleavedChangeColumnSize: {Brief: "Candidate for Interleaved Table", severity: suggestion, Category: "CHANGE_INTERLEAVED_COLUMN_SIZE",
+		CategoryDescription: "If column size of this table's primary key is changed to match the foreign key, the table can be interleaved"},
+	internal.RowLimitExceeded: {Brief: "Non key columns exceed the spanner limit of 1600 MB. Please modify the column sizes", severity: errors, Category: "ROW_LIMIT_EXCEEDED"},
+	internal.ShardIdColumnAdded: {Brief: "column was added because this is a sharded migration and this column cannot be dropped", severity: note, Category: "SHARD_ID_COLUMN_ADDED",
+		CategoryDescription: "Some Column was added because this is a sharded migration and that column couldn't be dropped"},
```
Comment (shreyakhajanchi) [inline internal/reports/report_helpers.go:551]: Shard id column was added because this is a sharded migration and this column cannot be dropped

File: internal/reports/report_helpers.go [hunk 22]
```diff
@@ -383,40 +507,53 @@ func fillRowStats(conv *internal.Conv, srcTable string, badWrites map[string]int
...
+	internal.ShardIdColumnPrimaryKey: {Brief: "column is not a part of primary key. You may go to the Primary Key tab and add this column as a part of Primary Key", severity: suggestion, Category: "SHARD_ID_ADD_COLUMN_PRIMARY_KEY",
+		CategoryDescription: "Some column is not a part of primary key. Check for that column and add it as a part of Primary Key"},
```
Comment (shreyakhajanchi) [inline internal/reports/report_helpers.go:553]: Shard id column is not a part of primary key. Please add it to primary key.

**Review by shreyakhajanchi — COMMENTED** [review]
Overall LGTM. Left minor comments

File: internal/convert.go [hunk 23]
```diff
@@ -410,6 +425,21 @@ func (conv *Conv) AddPrimaryKeys() {
 				ct.ColDefs[columnId] = ddl.ColumnDef{Name: k, Id: columnId, T: ddl.Type{Name: ddl.String, Len: 50}}
 				ct.PrimaryKeys = []ddl.IndexKey{{ColId: columnId, Order: 1}}
 				conv.SyntheticPKeys[t] = SyntheticPKey{columnId, 0}
+
+				// Adding 'Missing Primary Key' as a Warning inside ColumnLevelIssues of conv object
+				tableLevelIssues := conv.SchemaIssues[ct.Id].TableLevelIssues
+				columnLevelIssues := make(map[string][]SchemaIssue)
```
[the remark this answers is not in the document]
Comment (priyaannu) [reply internal/convert.go:431]: I did it to avoid accessing any null keys within the map at any time, however we can modify the code with better coding conventions. will update it

File: test_data/mysql_structured_report.json [hunk 24]
```diff
@@ -1,151 +1,172 @@
 {
-   "summary":{
-      "text":"Schema conversion: POOR (65% of 17006 columns mapped cleanly) + some missing primary keys.\nData conversion: POOR (66% of 6000 rows written to Spanner).\n",
-      "rating":"POOR",
+   "summary": {
+      "text": "Schema conversion: POOR (29% of 17006 columns mapped cleanly) + some missing primary keys.\nData conversion: POOR (66% of 6000 rows written to Spanner).\n",
```
Comment (darshan-sj) [inline test_data/mysql_structured_report.json:3]: What changed in the complexity logic? Why did 65% change to 29%?

File: test_data/mysql_text_report.txt [hunk 25]
```diff
@@ -1,7 +1,7 @@
 ----------------------------
 Summary of Conversion
 ----------------------------
-Schema conversion: POOR (65% of 17006 columns mapped cleanly) + some missing primary keys.
+Schema conversion: POOR (29% of 17006 columns mapped cleanly) + some missing primary keys.
```
Comment (darshan-sj) [inline test_data/mysql_text_report.txt:4]: Same comment. What changed in the complexity calculation logic?

File: ui/src/app/components/sidenav-view-assessment/sidenav-view-assessment.component.ts [hunk 26]
```diff
@@ -134,4 +160,144 @@ export class SidenavViewAssessmentComponent implements OnInit {
...
+        description: fetchedDescription,
+        tableCount: value.tableCount,
+        tableNamesJoinedByComma: tableNamesList.join(', '),
+      })
+      i += 1;
+    }
+  }
+
+  AppendIssueWithTableInformation(Issues: IIssue[], report: Map<string, TablesInformation>, defaultIssue: TablesInformation, fetchedTableReport: ITableReport) {
+    for (var noteIssue of Issues) {
```
Comment (darshan-sj) [inline ui/src/app/components/sidenav-view-assessment/sidenav-view-assessment.component.ts:278]: Is this method only for noteIssues?

File: webv2/config.json [hunk 27]
```diff
@@ -1,4 +0,0 @@
-{
- "GCPProjectID": "",
```
Comment (darshan-sj) [inline webv2/config.json:2]: Why is this deleted?

**Review by darshan-sj — COMMENTED** [review]
LGTM with minor comments

File: webv2/config.json [hunk 28]
```diff
@@ -1,4 +1,4 @@
 {
- "GCPProjectID": "",
- "SpannerInstanceID": ""
+ "GCPProjectID": "span-cloud-testing",
```
[the remark this answers is not in the document]
Comment (shreyakhajanchi) [reply webv2/config.json]: I didn't mean for you to delete the file, just to remove your settings

File: test_data/mysql_structured_report.json (hunk 24 above)
Comment (priyaannu) [reply test_data/mysql_structured_report.json:3]: Previously missing primary key was not included in the calculation of total number of warnings but now they are included and this impacted the rating of the schema

File: ui/src/app/components/sidenav-view-assessment/sidenav-view-assessment.component.ts (hunk 26 above)
Comment (priyaannu) [reply ui/src/app/components/sidenav-view-assessment/sidenav-view-assessment.component.ts:278]: Sorry this was a mistake, corrected it

File: webv2/config.json (hunk 27 above)
Comment (priyaannu) [reply webv2/config.json:2]: My bad! will add it back

File: webv2/config.json (hunk 28 above)
[the remark this answers is not in the document]
Comment (priyaannu) [reply webv2/config.json]: sure, my bad

File: test_data/mysql_text_report.txt (hunk 25 above)
Comment (priyaannu) [reply test_data/mysql_text_report.txt:4]: Previously missing primary key was not included in the calculation of total number of warnings but now they are included and this impacted the rating of the schema

**Review by shreyakhajanchi — APPROVED** [review]
LGTM
